.github/workflows/build-engine-branch.yml

```diff
@@ -0,0 +1,98 @@
+name: Build custom engine branch if needed
+on:
+  workflow_call:
+    outputs:
+      branchName:
+        description: The name of the branch of checked out engine
+        value: ${{ jobs.parseCommand.outputs.branchName }}
+      engineHash:
+        description: Commit hash of the checked out engine
+        value: ${{ jobs.build.outputs.engineHash }}
+
+jobs:
+  parseCommand:
+    name: Parse "/engine-branch" command in PR body
+    runs-on: ubuntu-latest
+    outputs:
+      branchName: ${{ steps.detectCustomEngineBranch.outputs.branchName }}
+    steps:
+      - name: Extract custom engine branch name from PR body
+        id: detectCustomEngineBranch
+        if: |
+          github.event_name == 'pull_request' &&
+          contains(github.event.pull_request.body, '/engine-branch') &&
+          (
+            github.event.pull_request.author_association == 'OWNER' ||
+            github.event.pull_request.author_association == 'MEMBER' ||
+            github.event.pull_request.author_association == 'COLLABORATOR' ||
+            github.event.pull_request.author_association == 'CONTRIBUTOR'
+          )
+        env:
+          PR_BODY: ${{ github.event.pull_request.body }}
+        run: |
+          BRANCH_NAME=$(echo "$PR_BODY" | grep -o '/engine-branch \S*' | awk '{print $2}')
+          if [ -z "$BRANCH_NAME" ]; then
+            echo "Branch name not specified in /engine-branch command."
+            exit 1
+          fi
+
+          echo "Requested custom engine from branch: $BRANCH_NAME"
+          echo "branchName=$BRANCH_NAME" >> $GITHUB_OUTPUT
+
+  build:
+    name: Build
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-13, windows-latest]
+    runs-on: ${{ matrix.os }}
+    needs: parseCommand
+    if: ${{ needs.parseCommand.outputs.branchName != '' }}
+    outputs:
+      engineHash: ${{ steps.engine-hash.outputs.engineHash }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          repository: prisma/prisma-engines
+          ref: ${{ needs.parseCommand.outputs.branchName }}
+      - name: Get current engine hash
+        id: engine-hash
+        shell: bash
+        run: echo "engineHash=$(git rev-parse HEAD)" >> $GITHUB_OUTPUT
+
+      - uses: actions/cache@v3
+        name: Artifacts cache
+        id: artifacts-cache
+        with:
+          path: |
+            target/release/schema-engine
+            target/release/schema-engine.exe
+            target/release/query-engine
+            target/release/query-engine.exe
+            target/release/libquery_engine.so
+            target/release/libquery_engine.dylib
+            target/release/query_engine.dll
+          key: ${{ runner.os }}-engine-${{ steps.engine-hash.outputs.engineHash }}-${{ github.event.number }}
```

Comment:
If two different PRs need the same version of the engine and this was built already, what's the reason to not reuse it? In other words, why adding `github.event.number`  to the cache key?

Reply:
My only thinking for that case was that in the case they are isolated by the PR, it makes it easier to find where and how the engine was actually built - it is always one of the commits on the same pull request. Where in other case, it might be actually impossible to figure out where we are getting the cache from without additional work. Reusing the cache will save us some time though, so in the end I don't have strong opinion. If you prefer to reuse the cache instead I can change it.

Reply:
No, your reasoning is perfectly valid, I think it's fine as it is, and then we can reevaluate whether we build the same engine from different PRs often. (I don't think is as likely to happen)

Thanks for explaining.